src/config/session.c

```diff
@@ -78,16 +78,51 @@ build_path(const char *dir, const char *filename)
 		return NULL;
 	}
 	int len = strlen(dir) + strlen(filename) + 2;
-	char *buffer = calloc(len, 1);
+	char *buffer = calloc(1, len);
 	strcat(buffer, dir);
 	strcat(buffer, "/");
 	strcat(buffer, filename);
 	return buffer;
 }
 
+static void
+update_activation_env(const char *env_keys)
+{
+	if (!getenv("DBUS_SESSION_BUS_ADDRESS")) {
+		/* Prevent accidentally auto-launching a dbus session */
+		wlr_log(WLR_INFO, "Not updating dbus execution environment: "
+			"DBUS_SESSION_BUS_ADDRESS not set");
+		return;
+	}
+	wlr_log(WLR_INFO, "Updating dbus execution environment");
+
+	char *cmd;
+	const char *dbus = "dbus-update-activation-environment ";
+	const char *systemd = "systemctl --user import-environment ";
+
+	cmd = calloc(1, strlen(dbus) + strlen(env_keys) + 1);
+	strcat(cmd, dbus);
```

Comment:
Basically you are reimplementing `asprintf()` here.

Reply:
> Basically you are reimplementing `asprintf()`

Yes, I am aware of this but:

> These functions are GNU extensions, not in C or POSIX.

> Also not checking for error code on calloc() is bad.

Agree'd. will change.

Reply:
Changed for the whole file.

Reply:
Yeah, but this function is present on all BSDs as well.

Reply:
But might not be in musl or other implementations, we try to not use any custom compiler extensions if at all possible.